dot/types/babe_digest.go

```diff
@@ -17,286 +17,179 @@
 package types
 
 import (
-	"encoding/binary"
 	"errors"
-	"io"
 
-	"github.com/ChainSafe/gossamer/lib/common"
 	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
 var _ BabePreRuntimeDigest = &BabePrimaryPreDigest{}
 var _ BabePreRuntimeDigest = &BabeSecondaryPlainPreDigest{}
 
 // BabePreRuntimeDigest must be implemented by all BABE pre-runtime digest types
 type BabePreRuntimeDigest interface {
-	Type() byte
-	Encode() []byte
-	Decode(r io.Reader) error
-	AuthorityIndex() uint32
-	SlotNumber() uint64
+	GetAuthorityIndex() uint32
+	GetSlotNumber() uint64
 }
 
-//nolint
-var (
-	BabePrimaryPreDigestType        = byte(1)
-	BabeSecondaryPlainPreDigestType = byte(2)
-	BabeSecondaryVRFPreDigestType   = byte(3)
-)
+// NewBabeDigest returns a new VaryingDataType to represent a BabeDigest
+func NewBabeDigest() scale.VaryingDataType {
+	return scale.MustNewVaryingDataType(BabePrimaryPreDigest{}, BabeSecondaryPlainPreDigest{}, BabeSecondaryVRFPreDigest{})
+}
 
 // DecodeBabePreDigest decodes the input into a BabePreRuntimeDigest
-func DecodeBabePreDigest(r io.Reader) (BabePreRuntimeDigest, error) {
-	typ, err := common.ReadByte(r)
+func DecodeBabePreDigest(in []byte) (BabePreRuntimeDigest, error) {
+	babeDigest := NewBabeDigest()
+	err := scale.Unmarshal(in, &babeDigest)
 	if err != nil {
 		return nil, err
 	}
 
-	switch typ {
-	case BabePrimaryPreDigestType:
-		d := new(BabePrimaryPreDigest)
-		return d, d.Decode(r)
-	case BabeSecondaryPlainPreDigestType:
-		d := new(BabeSecondaryPlainPreDigest)
-		return d, d.Decode(r)
-	case BabeSecondaryVRFPreDigestType:
-		d := new(BabeSecondaryVRFPreDigest)
-		return d, d.Decode(r)
+	switch msg := babeDigest.Value().(type) {
+	case BabePrimaryPreDigest:
+		return msg, nil
+	case BabeSecondaryPlainPreDigest:
+		return msg, nil
+	case BabeSecondaryVRFPreDigest:
+		return msg, nil
```

Comment:
```suggestion
	switch msg := babeDigest.Value().(type) {
	case BabePrimaryPreDigest, BabeSecondaryPlainPreDigest, BabeSecondaryVRFPreDigest:
		return msg, nil:
	default:
	         // never gonna happen right?
```

Reply:
This doesn't seem to work. The receiver functions for the types aren't found when I do this, so I think it needs to know the exact type being returned?